src/Interpreter/Mulang.hs

```diff
@@ -0,0 +1,387 @@
+  raiseCallback <- gets currentRaiseCallback
+  modify' (\c -> c {currentException = Just exceptionRef})
+  raiseCallback exceptionRef
+  raiseString "Unreachable" -- the callback above should never allow this to execute
+
+raiseString :: String -> Executable a
+raiseString s = do
+  raiseInternal =<< (createReference $ MuString s)
+
+muValuesEqual r1 r2
+  | r1 == r2 = createReference $ MuBool True
+  | otherwise = do
+      v1 <- dereference r1
+      v2 <- dereference r2
+      createReference $ MuBool $ muEquals v1 v2
+
+muEquals (MuBool b1)   (MuBool b2)   = b1 == b2
+muEquals (MuNumber n1) (MuNumber n2) = n1 == n2
+muEquals (MuString s1) (MuString s2) = s1 == s2
+muEquals MuNull        MuNull        = True
+muEquals _             _             = False
+
+getParamNames :: [Mu.Pattern] -> [String]
+getParamNames params =
+  fmap (\(Mu.VariablePattern n) -> n) params
+
+runFunction :: [Reference] -> Mu.Expression -> Executable Reference
+runFunction functionEnv body = do
+  context <- get
+  returnValue <- callCC $ \(returnCallback) -> do
+    put (context { scopes = functionEnv, currentReturnCallback = returnCallback })
+    evalExpr body
+    return nullRef
+  modify' (\c -> c { scopes = scopes context
+                   , currentReturnCallback = currentReturnCallback context
+                   , currentRaiseCallback = currentRaiseCallback context
+                   , currentException = currentException context
+                   })
+  return returnValue
+
+findFrameForName :: String -> Executable Reference
+findFrameForName name = do
+  maybe (raiseString $ "Reference not found for name '" ++ name ++ "'") return
+    =<< findFrameForName' name
+
+findFrameForName' :: String -> Executable (Maybe Reference)
+findFrameForName' name = do
+  framesRefs <- gets scopes
+  frames :: [(Reference, Map String Reference)] <- forM framesRefs $ \ref -> do
+    dereference ref >>= \value -> case value of
+      (MuObject context) -> return (ref, context)
+      v -> error $ "Finding '" ++ name ++ "' the frame I got a non object " ++ show v
+
+  return $ fmap fst . find (Map.member name . snd) $ frames
+
+
+findReferenceForName :: String -> Executable Reference
+findReferenceForName name = do
+  ref <- findFrameForName name
+  (MuObject context) <- dereference ref
+  return $ context Map.! name
+
+getMaxKey :: Map k a -> Maybe k
+getMaxKey m = case Map.maxViewWithKey m of
+  Just ((k, _a), _) -> Just k
+  _ -> Nothing
+
+nullRef = Reference 0
+
+dereference' :: ObjectSpace -> Reference -> Value
+dereference' _ (Reference 0) = MuNull
+dereference' objectSpace ref = do
+  fromMaybe (error $ "Failed to find ref " ++ show ref ++ " in " ++ show objectSpace) .
+    Map.lookup ref $
+    objectSpace
+
+dereference :: Reference -> Executable Value
+dereference ref = do
+  objectSpace <- gets globalObjects
+  return $ dereference' objectSpace ref
+
+updateGlobalObjects f context =
+  context { globalObjects = f $ globalObjects context
+          }
+
```

Comment:
I'd recommend one-lining simple expressions that don't require multiple span - `do`-expressions or really long applications. 

I particularly enforce this with solo-chars: 

```haskell
updateGlobalObjects f context =
  context { globalObjects = f $ globalObjects context }

updateGlobalObjects f context = context { globalObjects = f $ globalObjects context }
```